Tag design changes

frankieroberto opened the pull request:
This updates the design of the Tag component.

Design changes:

* no longer uppercase
* no longer bold
* remove 1px extra letter spacing
* increase font size to 19px
* change colours to use lighter background colours

These changes are to improve the accessibility and readability of the tag text, and to make the tags look less like clickable buttons.

These changes were originally designed as part of the 'status' component of the upcoming Task list component (#2261) – however feedback from the working group suggested these changes should also be applied to the Tag component for consistency.

querkmachine commented:
My quick 2 pence:

I'm not a fan. It looks too... unintentional, for want of a better phrase. 

Or, I guess, it doesn't look like it's a component that exists to serve a specific, unique purpose—it just looks like regular body text that's had a background colour added to it.

Some ideas, presented in isolation but could probably be combined if it's useful.

|Mockup|Description|
|:-|:-|
|![bold](https://user-images.githubusercontent.com/1253214/233611422-072a3672-3152-4b6e-be33-0f5d78aadba5.png)|**Increase contrast.** At the moment the text just looks a bit 'lost' in the background colour. Increasing the stroke width might help with that. Alternatively, tweaking the colours to increase the contrast between text and background might work, though we'd then risk the background becoming undiscernable to users with poor vision.|
|![border](https://user-images.githubusercontent.com/1253214/233611428-9b3f2977-de9d-4553-a5f6-32c4b281449c.png)|**Adding a border.** Adding a border (potentially in the same colour as the text) would make it feel a bit more distinct and intentioned. Mainly taking inspiration from [the NHS Design System's equivalent](https://service-manual.nhs.uk/design-system/components/tag) here.|
|![pills](https://user-images.githubusercontent.com/1253214/233611433-ed5b0bef-48f2-4dd1-b1c3-a4cac837a8fe.png)|**Adding border radius.** Rounding off the corners to make a 'pill' shape would certainly make it distinct, and create a resemblence to how tags and status indicators in other UIs are often presented (e.g. notification badges).|
|![dots](https://user-images.githubusercontent.com/1253214/233611431-8c054713-cd21-4553-8d85-4332bd1d990a.png)|**Something else entirely?** The [ONS Design System](https://service-manual.ons.gov.uk/design-system/components/status) uses coloured dots and plain text to indicate status, rather than changing the colour and background of the text itself.|

Admittedly many of these start veering back into the realm of 'could be a button', but I also feel that it's kind of impossible to put a box around some text and _not_ have someone think it's clickable. I think our goal is to distinguish them from the few existing buttons found in the Design System, rather than try and divorce them from the concept of buttons entirely.

frankieroberto commented:
> Or, I guess, it doesn't look like it's a component that exists to serve a specific, unique purpose—it just looks like regular body text that's had a background colour added to it.

For the task list, that was arguably the intention. HMRC decided to do away with the tag style altogether and [just use regular text](https://design.tax.service.gov.uk/hmrc-design-patterns/status-tags-in-task-list-pages/):

<img width="611" alt="Screenshot 2023-04-21 at 12 04 14" src="https://user-images.githubusercontent.com/30665/233620749-58a25909-2a45-4878-8f5c-16ce1de94f1c.png">

Within the task list working group we reached a position of using regular text for completed tasks, and "highlighted" text for tasks yet to be completed, in order to draw more attention to those.

There was a debate about whether the task list statuses needed to use the same style as the Tag component at all. We initially deviated from the tag component styles, but the Design System working group thought they should be consistent.

I’m open to all ideas – although the Design System team get the final say!

Would be good to come to a resolution fairly soon though, as this is blocking the release of the task list component.

paulrobertlloyd commented:
That’s the word @beeps: *intention*. The current proposal feels too unintentional, almost like a bit of CSS hasn’t loaded.

As to the other ideas:

* **Increase contrast:** I agree, I think the text still feels a bit washed out. Not sure if the text needs to be emboldened, but if there was an additional weight in the font palette between regular and bold, I’d be suggesting that. Given there isn’t, it's a case of having to choose between the 2 weights.
* **Adding a border:** This might be an option. Make the border too dark though, and it adds visual noise, and can affect layouts that already have a lot of borders and boxes, leading to cases of lots of nested boxes. I think I’d be playing around with options that were so faint as not to bother.
* **Adding a border radius:** I wondered about this as well, but GOV.UK isn’t a round corners site. That said, the icons added to the revised accordion component start to go a bot in that direction, and as I mentioned in relation to that, there *are* circles in the GOV.UK Crown logo.
* **Something else:** I’ve in fact created separate components in the past with dots to the left of text, but that was more to indicate a status, whereas the tag component is used more broadly.

Which brings us on to the Phase banner. Hmmmmm, this is tricky, and starts to pull on a horrible thread of consequences. The font-size mismatch gets tricky. Options:

* Create a small variation of the tag component (which I’ve done on a number of occasions in the past anyway) and use that – not unprecedented either as radios/checkboxes have small variation
* Try and massage the tag component so that it works in all these different contexts, and not be too concerned that text isn’t aligned to baselines and appears at different sizes
* Use a bespoke design for the phase banner… and congratulations you’re now redesigning three different components!

Think I’d be tempted to go for option 1.

I think my changes would be:

* adjust padding on tag component to make it feel more intentionally designed, and look less like highlighted text
* increase text contrast on tag component a tad more
* add a small variation to the tag component, and use that in the phase banner component

paulrobertlloyd commented:
Looking back at my earlier suggestion for padding, I think the values lean possibly to far towards much on the large (and possibly button-y) side. Have revised it to use block padding of 3px (2 less than 5), inline padding of 8px (2 less than 10) and top/bottom margin of -3px:

![Screenshot 2023-04-21 at 13 42 07](https://user-images.githubusercontent.com/813383/233639144-68d4c9b0-cb86-477c-9d47-0a22e345d2f1.png)

steve-oconnor commented:
@querkmachine We did go over most - if not all - of those particular ideas, but we had specific constraints at the time and what we ended up with was voted the best option. The constraints may have shifted a little now.

All colours where contrast checked (when I did them anyway!) so that they met AA+ standards.

@paulrobertlloyd I personally prefer your second design - less buttony! Keeping them looking like highlighted text and not like buttons (as current tags do, according to user research) was a deliberate design aesthetic.

steve-oconnor commented:
I'm not sure where all the extra colour options came from either! 4 pairs of the designs being shared look almost identical. The original idea was to have a limited palette as each task list design we saw had a limited set of tags. If a particular team wanted to extend the palette, they could - with testing to back it up.

Each colour represents a *type* of tag, not a specific tag name. This should be written up in the guidance.

The green tag on one task list might be "New", on another it might be "Active". Blue might be "In progress" on one task list, "Pending" on another. Most users will not be using all the possible task lists, and where there are multiple task lists that the same people might access, that particular set should follow the same pairings pf colour and words.

querkmachine commented:
@steve-oconnor Thanks for the context! This is the first time I'm seeing this, personally, so I'm not clued up on any prior considerations. 

The extra colour options are to reflect the options already present in the Design System, viewable [on the Tag documentation page](https://design-system.service.gov.uk/components/tag/#additional-colours). If we're changing the design of the component, we probably want to keep accommodating all of the existing colour options, as they might've been used by a team somewhere already.

frankieroberto commented:
> I'm not sure where all the extra colour options came from either! 

As @querkmachine corrected guessed, there are already a bunch of different tag colours, so when I copied over the colours you did for the task list @steve-oconnor, I just reused the existing colours, with the exception of "light blue" (which had to be added), and the "text colour" and "secondary text colour" ones.

Both the existing tag guidance, and the [draft task list component guidance](https://deploy-preview-1994--govuk-design-system-preview.netlify.app/components/task-list/) suggest using as few different statuses and colours as possible, so hopefully that would dissuade teams from using all of them at once...

paulrobertlloyd commented:
This PR also adds 2 new ‘tags’, which appear heavily tied to the implementation for the task list component, without any clear indication how they might be used outside of it:

* `govuk-tag--text-colour`
* `govuk-tag--secondary-text-colour`

Perhaps this verges too far into discussion about the task list component, but do these exist because that component expects each status to use a tag? Where might you use these variations outside the confines of the task list component?

(FWIW, a broader application utility class like `govuk-!-secondary-text-colour` would be massively helpful, and perhaps such an addition would aid in the design/development of the task list component).

owenatgov reviewed (commented):
This looks technically sound to me 👍🏻 Since it doesn't impact the component API or markup in a destructive way and isn't a design change that could disrupt any design structures, visual or otherwise, that this doesn't need to be a breaking change and can go into a minor (0.X.0) change at a minimum.

Being pedantic, I would like us to clear up the commit history. Probably it can be squashed into one commit. I am in particular keen for changes to be reflected in tests in the same commit if we can help it.

christopherthomasdesign commented:
Yeah I think Paul's second set looks good @frankieroberto I think those make them look a bit more solid – want to apply those?

frankieroberto commented:
@christopherthomasdesign @owenatgov ok I've updated the pull request with the extra padding from @paulrobertlloyd’s design, fixed the conflicts with `main` and squashed all the commits.

I've also done another pre-release so that this preview for the PR on the design system website repo has been updated: https://deploy-preview-2733--govuk-design-system-preview.netlify.app/components/tag/

Let me know if there’s anything else to do!

paulrobertlloyd commented:
Ah, nice! The new tag design will take a bit getting used to, but overall a big win for legibility and accessibility. 🚀 

I’d still like to challenge the inclusion of the `govuk-tag--text-colour` and `govuk-tag--secondary-text-colour` variants as they are coupled to design of the task list component and I can’t see a use case for these variants outside of that.

I think there could be some better, more broadly scalable options for addressing the styling of ‘non-coloured’ statuses in that component, rather than depending on these 2 specific tag variants (which I described in the discussion thread for that component).

frankieroberto commented:
> Great work folks! As a prompt, it may we worth checking how this looks against [the new typography scale](https://github.com/alphagov/govuk-frontend/pull/2421). 

I've had a quick check, and I think it should be fine. It may mean more wrapping of text within the Task list on narrower screens, but the design can cope with that anyway (and it's a valid trade-off for the benefit of larger text).

frankieroberto commented:
@steve-oconnor:

> The only thing I'm still not sold on is the use of the grey "inactive" tag rather than the text with no highlight. The original intention was to have guidance to explain how coloured tags should be used to highlight items that need attention, plain text for items that don't require attention.

This is all the existing [guidance for the Tag component](https://design-system.service.gov.uk/components/tag/), which I’ve not otherwise reviewed. I think those examples are for where the tag is used outside of the task list (eg in in table), and I’m not sure that the text or colours are intended to be prescriptive? However it may still be worth reviewing separately?

For the [task list guidance](https://deploy-preview-1994--govuk-design-system-preview.netlify.app/components/task-list/) we’re being a bit more prescriptive (or at least suggestive) of how to pair the colours and text - with completed tasks using the regular black text, as you suggest.

frankieroberto commented:
@paulrobertlloyd:

> I’d still like to challenge the inclusion of the `govuk-tag--text-colour` and `govuk-tag--secondary-text-colour` variants however, certainly at this stage. These are coupled to design of the task list component and I can’t see a use case for these variants outside of that.
> 
> I think there could be some better, more broadly scalable options for addressing the styling of ‘non-coloured’ statuses in that component, rather than depending on these 2 specific tag variants (which I described in the discussion thread for that component).

I think there _could_ be uses for these outside of the task list. Whilst you could just use regular text for the `govuk-tag--text-colour` variant, using the Tag component with the class will also add the extra padding (that you suggested) which will mean it'll line up nicely with any coloured variants.

A more generic override class like `govuk-!-text-colour` could be an option, but it would also need to reset the background colour back to transparent so you'd need something like `govuk-!-background-colour-transparent`, and at that point you may as well define a tag-specific modifier class?

I've now updated the task list component in https://github.com/alphagov/govuk-frontend/commit/ba1e20e7ebe4adfd6b815d437b0113aac3680bd7 so that it directly uses the Tag component – which means there’s no easy way to avoid using the Tag altogether and define your own html within the tag container (except by using HTML and not the Nunjucks macro).

paulrobertlloyd commented:
This verges into discussion about the task list component, but…

> Whilst you could just use regular text for the govuk-tag--text-colour variant, using the Tag component with the class will also add the extra padding (that you suggested) which will mean it'll line up nicely with any coloured variants.

I’m not sure you need the extra padding for the text coloured statuses.

> A more generic override class like govuk-!-text-colour could be an option, but it would also need to reset the background colour back to transparent so you'd need something like govuk-!-background-colour-transparent, and at that point you may as well define a tag-specific modifier class?

No modifier, no background colour, just use text as is used in other component.

> I've now updated the task list component in https://github.com/alphagov/govuk-frontend/commit/ba1e20e7ebe4adfd6b815d437b0113aac3680bd7 so that it directly uses the Tag component – which means there’s no easy way to avoid using the Tag altogether and define your own html within the tag container

Not sure if you’ve seen my suggestion, but my thought was you could either provide text or a tag (`status.text` or `status.tag`). That saves having to think of every status requiring an equivalent tag modifier.

* * *

I’d recommend keeping this change scoped to the visual design of the existing tag component and colour modifiers. If it’s decided that further tag modifiers are needed after review of the task list component, those could be added as part of that change. Otherwise the changes (and discussion) gets a bit muddled.

christopherthomasdesign commented:
One stylistic tweak I should have raised earlier – we usually try to stick to a 5px vertical rhythm for components and typography (hence why all our line heights are in multiples of 5px etc). It's not applied 100% consistently but I do think it helps keeping everything looking broadly proportionate. Do you think we lost anything by shaving a pixel off the padding-top of the tags, so they're 30px heigh rather than 31? Screenshot below:

<img width="778" alt="Screenshot 2023-05-04 at 15 43 54" src="https://user-images.githubusercontent.com/19834460/236243641-fbd4ba53-ca13-4722-8218-ddaa689081e5.png">

I think I need a bit more time to work out what I think about the different tag variants and how much they should overlap with the task list, bear with me on that one...

frankieroberto commented:
@paulrobertlloyd:

> I’m not sure you need the extra padding for the text coloured statuses.

Hmm maybe you’re right. I’m not sure which I prefer. Here’s a quick mock-up showing the "Completed" statuses without the extra 8px padding on the right.

<img width="780" alt="Screenshot 2023-05-04 at 16 27 39" src="https://user-images.githubusercontent.com/30665/236255276-e25c533c-b582-42f4-918b-1f336eae5a97.png">

vs with the padding:

<img width="792" alt="Screenshot 2023-05-04 at 16 27 29" src="https://user-images.githubusercontent.com/30665/236255219-d701a382-2b37-4a15-8be8-d1299e03569f.png">

frankieroberto commented:
@paulrobertlloyd:
> Not sure if you’ve seen my suggestion, but my thought was you could either provide text or a tag (status.text or status.tag). That saves having to think of every status requiring an equivalent tag modifier.

Could be an option - but it does make the nunjucks macro slightly more complicated/nested (from `items.tag.text` to `items.status.tag.text`)..?  On the other hand it would let you add classes to both the tag and the container element?

christopherthomasdesign commented:
Good spot, cheers @36degrees – I personally would just bump up the text in the phase banner but am happy to be challenged on that, it might need a bit of testing first.

On the capital letter thing, I did discuss this with @owenatgov the other day – wonder if you had any thoughts on how to handle this?

frankieroberto commented:
@36degrees:

> The baseline on the [phase banner](https://deploy-preview-2733--govuk-design-system-preview.netlify.app/components/phase-banner/) is a little off with these changes. We might also want to consider making the phase banner text 19px as well – at the minute it just looks off to me.
> 
> <img alt="Screenshot 2023-05-04 at 16 31 46" width="783" src="https://user-images.githubusercontent.com/121939/236256889-5d8de194-b06d-4418-b939-ed227ce00d4c.png">
> 
> Alternatively, we might consider keeping the existing design for the phase banner, removing its dependency on the tag component.

@christopherthomasdesign and I discussed this briefly. Bumping up the text in the phase banner feels like it'd be the least disruptive (no breaking changes).

A more radical option might be to break the dependency on Tag and just use some plain bold text, or similar?

![Screenshot 2023-04-21 at 13 54 39 (2)](https://user-images.githubusercontent.com/30665/236262840-e79a1e46-e74f-4c43-9ff9-7daebd9440e0.png)

> Secondly, at the minute the changes require teams to update their markup to uppercase the first letter. For example, from `alpha` to `Alpha`. I think we’re at risk of seeing tags (especially in phase banners) in lowercase all over the place if we’re not careful.
> 
> <img alt="Screenshot 2023-05-04 at 16 35 26" width="783" src="https://user-images.githubusercontent.com/121939/236257170-652b6d6c-fd69-4b3e-a68c-a2c602a35a07.png">
> 
> Hardly world-ending, but we might want to consider doing something to try and avoid this? Could look at `text-transform: capitalize` (though that would capitalise every word in a tag, e.g. 'In Progress', 'Cannot Start Yet') or doing something clever with `::first-letter`.

I think I'd probably try and avoid doing something clever with CSS? (`text-transform: capitalize` is a non-starter). Hopefully teams would only have to change the capitalization in a single file – as services only include a single phase banner shared across every page? Would this still count as a breaking change though? 🤔

paulrobertlloyd commented:
> My vote would be for the second option, with padding. Ease of readability.

The designer in me does prefer it when things line up, the pragmatist less so 🙂 

In terms of implementation, not sure you need an additional tag modifier to achieve this. The status containing element could have `padding-right: 8px`, and a `.govuk-tag` within that could have `margin-right: -8px`. Doesn’t have to be that, I guess my broader point is that it seems weird to create a tag to achieve a component-specific spacing.

querkmachine commented:
Is future proofing an argument for a dedicated tag? 

As a hypothetical, maybe some users just don't consider the secondary text colour one to be visually distinct enough and we decide to change it. We could update the CSS for a dedicated tag without needing a breaking change, but potentially not if we used code with totally distinct HTML/class names.

colinrotherham commented:
@frankieroberto Sorry just being nosey since the big directory changes 🙈 

Can you double check you meant for both `src` and `dist` stylesheets to be updated? Only because we've seen **git** get confused by the two files starting off the same—then replaying changes on the wrong one:

```
packages/govuk-frontend/src/govuk/components/tag/_index.scss
packages/govuk-frontend/dist/govuk/components/tag/_index.scss
```

https://github.com/frankieroberto/govuk-frontend/commit/3f520ed6b5efc5da6c219ea9293f655c4d8ef784#diff-930b1115bd006cc9993120607535cd35b44dfa89171dfc77d5feee98601ebfbf

frankieroberto commented:
@colinrotherham:
> Can you double check you meant for both `src` and `dist` stylesheets to be updated? 

I did not - but rebasing from main since the big directory changes caused a few issues! 😬 Fixed manually now (and forced pushed another rebase).

frankieroberto commented:
@36degrees we chatted a bit about this on the Friday community call (in the context of “what makes a breaking change?”)

On reflection maybe we should just accept it as a breaking change - and this might then also free us up to make a breaking change to the tag within the phase banner?

The `::first-letter` css selector might still not work where services have words within tags which they’d still want capitalised - like two-word proper nouns, or abbreviations? Probably fairly rare, but I think tags do get used for all sorts of labels within case working services?

@christopherthomasdesign what do you think?

steve-oconnor commented:
> > My vote would be for the second option, with padding. Ease of readability.
> 
> The designer in me does prefer it when things line up, the pragmatist less so 🙂
> 
> In terms of implementation, not sure you need an additional tag modifier to achieve this. The status containing element could have `padding-right: 8px`, and a `.govuk-tag` within that could have `margin-right: -8px`. Doesn’t have to be that, I guess my broader point is that it seems weird to create a tag to achieve component-specific spacing.

It's the accessibility nerd in me 🤓 

I don't think it's weird to create a tag to achieve the spacing. It makes it known that this text is still a tag, semantically, and the style can be changed in the future without someone having to go back and check where it's been set as plain text in pages and needs to be manually updated.

christopherthomasdesign commented:
I've just been looking at how the tags would affect the phase banner in a bit more detail, it's proving to be a tricky dependency in terms of how the styles carry over.

Here's the new tag applied to the phase banner without changing anything else (except for the capitalisation, which I've changed for all these examples):

![Screenshot 2023-05-10 at 10 36 41](https://github.com/alphagov/govuk-frontend/assets/19834460/936370c4-85d1-4645-8e75-04c5a9203123)

As discussed before, I think the mismatch between the text sizes is a bit jarring here.

Here it is with the text bumped up to 19px to match the tags:

![Screenshot 2023-05-10 at 10 44 21](https://github.com/alphagov/govuk-frontend/assets/19834460/ea0dc272-5690-4bd1-8927-75b3120e9815)

I think this looks fine until you factor in things like breadcrumbs and back buttons sitting below. Once you have those, it feels like it messes with the page hierarchy a bit. To me the phase banner should probs be at the same level as those elements, where they're 'in the background' compared to the main page content.

Here's a version where I've reduced the text in the tag down to the 16px size in the font scale:

![Screenshot 2023-05-10 at 10 45 47](https://github.com/alphagov/govuk-frontend/assets/19834460/be344b1c-4f27-4830-a896-931257da242c)

That makes the tag a total of 25px height – that looks a bit small to me here but it's actually the same height it was before. 

I wonder if adding something custom to the phase banner to achieve this might be a way forward, though I I'll admit I've got a bit mixed up on what we're saying in terms of releases and breaking changes (@owenatgov we might need another chat about this!). I'm not sure if you'd need to break the existing phase banner html though.

As ever, thoughts welcome!

frankieroberto commented:
@christopherthomasdesign I’d agree that the version with the tag text reduced works best.

In terms of implementation, I guess we either do something where we use CSS to target tags within phase banners, which would be backwards-compatible, or we add a modifier class for the tag, but that'd require some html changes to add the extra class?

christopherthomasdesign commented:
> @christopherthomasdesign I’d agree that the version with the tag text reduced works best.
> 
> In terms of implementation, I guess we either do something where we use CSS to target tags within phase banners, which would be backwards-compatible, or we add a modifier class for the tag, but that'd require some html changes to add the extra class?

What do you reckon @owenatgov?

owenatgov commented:
My personal preference would be to target tags within phase banners via CSS. We do something similar with [summary lists within summary cards](https://github.com/alphagov/govuk-frontend/blob/main/packages/govuk-frontend/src/govuk/components/summary-list/_index.scss#L255).

frankieroberto commented:
@owenatgov ok, this was an easy fix: e3e67520ea3a1a1ec82f17c8ebd03a5262ff751f

frankieroberto commented:
@christopherthomasdesign @owenatgov first letter uppercased in 954a54fe7f1f23c2e6687407304f684850a323a8.

I think this is now good to merge, and no longer a breaking change?

colinrotherham commented on packages/govuk-frontend/src/govuk/components/tag/_index.scss:
```diff
@@ -54,13 +67,13 @@
-    color: govuk-shade(govuk-colour("pink"), 40%);
-    background: govuk-tint(govuk-colour("pink"), 80%);
+    color: govuk-shade(govuk-colour("pink"), 50%);
+    background-color: govuk-tint(govuk-colour("pink"), 85%);
```
Do you know why some colours use `background-color` and others `background`?

frankieroberto replied:
Well spotted! I think this was an existing inconsistency. I've changed them all to use `background-color` now in 2366183, thanks!

Ciandelle commented:
The use of either dashed or double lined boxes looks great to me, however the overall decision for the use case of these needs to be confirmed by the working group. 

On a separate point, as someone outside of the work, this definitely feels less like a button and the larger text helps to make it stand out. Excellent work folks!

owenatgov commented:
We're very very close. The code is looking fine by my eye.

@davidc-gds ran a quick test against forced colour mode, specifically windows high contrast, and found a final issue. The tag component now looks indistinguishable from our button in high contrast mode:

![image](https://github.com/alphagov/govuk-frontend/assets/64783893/670eca61-1d4b-4f57-a072-1f782875f408)

Our suggestion is to use a different [border style](https://developer.mozilla.org/en-US/docs/Web/CSS/border-style) such as dashed in forced colours mode:

![image (1)](https://github.com/alphagov/govuk-frontend/assets/64783893/090719dc-8435-4ff7-a7ed-ec7b353fa962)

You could do this quite easily using a forced colours media query:

```scss
@media screen and (forced-colors: active) {
  border-style: dashed;
}
```

paulrobertlloyd commented:
Is a border needed at all for high contrast mode? Given the [background colour should not be used alone to convey information](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html), how important is it that tag text is distinguishable from other text on a page?

frankieroberto commented:
I think I’d agree with @paulrobertlloyd that it'd be better to have no border in high-contrast mode rather than a dashed one (whose meaning might be hard to interpret), given that the tag text is just informational and not clickable in any way?

davidc-gds commented:
If the outline is fully removed, would this mean that for high contrast mode, and people with low-vision there are potentially no signifiers to differentiate the tag component from plain text?

frankieroberto commented:
> If the outline is fully removed, would this mean that for high contrast mode, and people with low-vision there are potentially no signifiers to differentiate the tag component from plain text?

Yes, other than placement (eg being aligned-right within the task list).

I guess the question is whether there's any need to distinguish it from plain text? (Which might be context-dependent?)

owenatgov commented:
As a go-between, could you make the text bold in forced colours mode? Similar to how we use `strong` to visually differentiate the tag from text in the absence of CSS.

That's if there's a need. My take is that there usually isn't an explicit need but it can't hurt to have a failsafe in place.

owenatgov commented:
@frankieroberto In that case I think the capitalisation alone would be enough but I also think the bold doesn't create any issues and even more clearly communicates the separation. I'm open to being disagreed with but that's my preference so far.

paulrobertlloyd commented on packages/govuk-frontend/src/govuk/components/tag/_index.scss:
```diff
@@ -1,12 +1,15 @@
     display: inline-block;
 
-    padding-top: 5px;
+    margin-top: -3px;
```
These margins are intended to be global so that using a tag doesn’t increase the height of a parent container. For example, if you have a table row with a tag, and one without, their heights will be different.

paulrobertlloyd replied:
Example:

![tag-spacing](https://github.com/alphagov/govuk-frontend/assets/813383/5cfc9d32-b981-4cee-98b6-8827da08b576)

Maybe we should add a comment to explain these negative margin values?

frankieroberto replied:
Added a comment in 68683bc – @paulrobertlloyd would you mind double-checking that the maths is right?

paulrobertlloyd replied:
Oh, actually the negative margins should match their equivalent positive padding values, meaning the top margin should be `-2px` (this doesn’t actually change the spacing, but it’s less clear why the value would be different).

paulrobertlloyd replied:
@frankieroberto Did you update this? /cc @owenatgov

frankieroberto replied:
@paulrobertlloyd I can't actually remember, there’s been so many tweaks and changes! 

But making the negative margin match the padding makes sense!

@colinrotherham I'm not familiar with how the `govuk-em` mixin works. Should we use this for all the `px` values?

colinrotherham replied:
Yes please to making them both match ✅

Regarding text scaling (e.g. Safari `⌥ ⌘ +` **Make Text Bigger**) I'm going to tag in @owenatgov

I'd like to say yes but I'm aware we have a [hard 10px spacing in `phase-banner`](https://github.com/frankieroberto/govuk-frontend/blob/552a49ad67308eb1feaea66eb32caef92d694dc1/packages/govuk-frontend/src/govuk/components/phase-banner/_index.scss#L5) too:

```css
padding-top: govuk-spacing(2);
padding-bottom: govuk-spacing(2);
```

So if we make the the tag itself "font-size scalable" the phase banner itself won't be

frankieroberto replied:
@colinrotherham I've made the negative margin match the padding in 54390308dc46d64203da7d76015db4571adc4811 (and added a comment to explain).

Do you want me to replace all the static `<x>px` values with `govuk-em(<x>, 16)` or should that be left to a follow-up PR? @owenatgov

paulrobertlloyd commented:
> This extra spacing is happening in the width of the tag element itself, not in the padding or margin according to firefox. Also according to firefox's accessibility tree, this is coming from the text, with the contents saying `"Beta "`.

Would adding whitespace control in the variable tag help?

```diff
  <strong class="govuk-tag{% if params.classes %} {{ params.classes }}{% endif %}"{% for attribute, value in params.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
-   {{ params.html | safe if params.html else params.text }}
+   {{- params.html | safe if params.html else params.text -}}
  </strong>
```

Or alternatively:

```njk
<strong class="govuk-tag{% if params.classes %} {{ params.classes }}{% endif %}"{% for attribute, value in params.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>{{ 
  params.html | safe if params.html else params.text
}}</strong>
```

owenatgov commented:
@paulrobertlloyd It's a good guess but sadly didn't make a difference for me.

From playing around, it looks like the issue is specifically `text-transform: uppercase` inside the `::first-letter` selector. I'm not really sure why firefox doesn't like this (could be a vendor bug?), but the fix I found was to replace it with `text-transform: capitalization`. I know this has been thrown out already, however localising it to `::first-letter` seems to solve the worries we had with it.

paulrobertlloyd commented:
> It's a good guess but sadly didn't make a difference for me.


![GIF of Roy Walker saying ”It’s good but it’s not right” from the gameshow ‘Catchphrase’.](https://c.tenor.com/mXlCpjAihJQAAAAC/catch-phrase-roywalker.gif)

frankieroberto commented:
> @paulrobertlloyd It's a good guess but sadly didn't make a difference for me.
> 
> From playing around, it looks like the issue is specifically `text-transform: uppercase` inside the `::first-letter` selector. I'm not really sure why firefox doesn't like this (could be a vendor bug?), but the fix I found was to replace it with `text-transform: capitalize`. I know this has been thrown out already, however localising it to `::first-letter` seems to solve the worries we had with it.

Wait, so `::first-letter { text-transform: uppercase; }` is rendering differently from `::first-letter { text-transform: capitalise; }`? That really does sound like a vendor bug?! 🤯

frankieroberto commented:
> @frankieroberto Yeah it's pretty whacky. I'll be raising a vendor bug and seeing what they say. We have a workaround, weird as it may be 😎

Fixed in b49afa3 – thanks!

frankieroberto commented:
@owenatgov so I think the border in high contrast mode was being added by the transparent outline previously added to make the Tag perceivable (see [blog post](https://accessibility.blog.gov.uk/2017/03/27/how-users-change-colours-on-websites/)). 

I’ve removed this in 903f575 (as it makes them indistinguishable from buttons) and have instead made the Tag text bold in high contrast mode.

Would you mind letting me know if this works, and if the comment to explain it is clear enough?

owenatgov commented:
We're mere steps from the finish line. 2 lingering things to do from me:

- When I test in windows high contrast, the bold isn't coming through. This is coming from the default font weight rule from `govuk-font` in `.govuk-phase-banner__content__tag` which has higher specificity. There are a couple of ways we could solve this:
    - Add `$weight: false` to the `govuk-font` declaration in `.govuk-phase-banner__content__tag`. This will override the default weight declaration when `govuk-font` is called. I've tested this and it works but it's a bit icky. If we go with this I'd like it explained with a comment.
    - Replicate the same forced colours media query in the phase banner content tag. This is nice and neat with the only downside being repeated code. If we need to do that we need to do that.
- I was thinking about how to segment the commit history and my personal take is either squash it all again or separate the tag redesign and the associated changes to the phase banner. Open to alternative perspectives.

frankieroberto commented:
@owenatgov 
> * When I test in windows high contrast, the bold isn't coming through. This is coming from the default font weight rule from `govuk-font` in `.govuk-phase-banner__content__tag` which has higher specificity. There are a couple of ways we could solve this...

I’ve gone with your second option (replacing the high-contrast mode css within the phase banner css). Whilst I personally like the idea of relying on cascading styles more, the conventions of BEM, and govuk-frontend more generally, has been to apply styles at the lowest level and avoid cascading? (I think this was to allow govuk-frontend to be adopted more incrementally alongside any existing css?)

> * I was thinking about how to segment the commit history and my personal take is either squash it all again or separate 
the tag redesign and the associated changes to the phase banner. Open to alternative perspectives.

I've squashed them all into a single commit again. 👍

owenatgov reviewed (approved):
Well done everyone 💪🏻

I'm going to ask @christopherthomasdesign to go over this one more time and either give a 2nd approval or comment if there's anything lingering. Pending that, I think we're ready to merge this!

colinrotherham commented on packages/govuk-frontend/src/govuk/components/phase-banner/_index.scss:
```diff
@@ -17,7 +17,16 @@
+
+    // When forced colour mode is active, for example to provide high contrast,
+    // the background colour of the tag is the same as the rest of the page. To ensure
+    // that that tag is perceived as separate the rest of the text in the phase banner,
```
Got a double "that that" and a missing word in this block

If you can fix this one and use the same comment further down too?

```suggestion
    // that tag is perceived as separate to the rest of the text in the phase banner,
```

frankieroberto commented on packages/govuk-frontend/src/govuk/components/tag/_index.scss:
```diff
@@ -1,70 +1,92 @@
+
+    // When forced colour mode is active, for example to provide high contrast,
+    // the background colour of the tag is the same as the rest of the page. To ensure
+    // that that tag is perceived as separate from any surround text, it is made bold.
```
@colinrotherham thanks, fixed the typo. 🤦   

I copied and pasted the comment, but tweaked it slightly to make it more specific to the phase banner. Sensible?

frankieroberto commented:
Re-squashed the commits and force-pushed. I think we’re there?? Anything needed for release notes etc, or do those only get written when preparing a release?

owenatgov commented:
@frankieroberto Good shout! Completely slipped my mind 🙃 I reckon let's do something in the changelog now for reference and we can refine this when we get into the release notes phase of 5.0. A heading saying that the design's been updated, noting that folk's spacing may move around and then link to the PR will do.

christopherthomasdesign reviewed (approved):
Have had a good nosey and am all happy with it from a design point of view 👍

frankieroberto commented:
@owenatgov:

> let's do something in the changelog now for reference and we can refine this when we get into the release notes phase of 5.0. A heading saying that the design's been updated, noting that folk's spacing may move around and then link to the PR will do.

Done in 625532f 👍

frankieroberto commented:
🏷️ 🏷️ 🏷️ 🏷️  Big thanks to @36degrees @christopherthomasdesign @colinrotherham @owenatgov @paulrobertlloyd @querkmachine @steve-oconnor! 🏷️ 🏷️ 🏷️ 🏷️

joelanman commented:
sorry to resurrect a very old thread, but is it intentional that the light grey is slightly different to the gov.uk light grey? I'm using tags and govuk light grey on a page, and they don't match
